### PR TITLE
Verify WFTs with large event buffers fail

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I added a unit test which verifies that we fail WFTs with event buffers that are too large.

**Why?**
To make it easier to add the event buffer byte size check to the CloseTransactionAsMutation.

**How did you test it?**
I verify that the WFT attempt was incremented and the WFT-failed event in the history exists and looks correct.

**Potential risks**
Just a test.

**Is hotfix candidate?**
No

### CODE REVIEW

File: service/history/workflow/workflow_test/mutable_state_impl_test.go [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func (c mutationTestCase) createMutableState(t *testing.T, nsEntry *namespace.Namespace) *workflow.MutableStateImpl {
+	t.Helper()
+
+	ctrl := gomock.NewController(t)
+	shardContext := shard.NewMockContext(ctrl)
```
Comment (yycptt) [inline service/history/workflow/workflow_test/mutable_state_impl_test.go:174]: there's a test implementation for shardContext you can use, and avoid some EXPECT() call like generating taskID or getting timeSource/metricsHandler/etc. .
check history/shard/context_testutil.go

File: service/history/workflow/workflow_test/mutable_state_impl_test.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/primitives/timestamp"
+	"go.temporal.io/server/service/history/configs"
+	"go.temporal.io/server/service/history/events"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/tests"
+	"go.temporal.io/server/service/history/workflow"
+)
+
+func TestMutableStateImpl_CloseTransactionAsMutation(t *testing.T) {
```
Comment (yycptt) [inline service/history/workflow/workflow_test/mutable_state_impl_test.go:59]: I would call this test TestMutableStateImpl_ForceFlushBufferedEvents.

File: service/history/workflow/workflow_test/mutable_state_impl_test.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+				WorkflowRunTimeout:  timestamp.DurationPtr(200 * time.Second),
+				WorkflowTaskTimeout: timestamp.DurationPtr(1 * time.Second),
+			},
+		},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	c.completeOneWFT(t, ms)
```
Comment (yycptt) [inline service/history/workflow/workflow_test/mutable_state_impl_test.go:123]: why do we need to complete one wft first?

File: service/history/workflow/workflow_test/mutable_state_impl_test.go [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+	if wftAttempt != 2 {
+		t.Errorf("Expected WFT attempt number to be 2 if the WFT failed, but was %d", wftAttempt)
+	}
+
+	event, ok := c.findWFTEvent(enumspb.EVENT_TYPE_WORKFLOW_TASK_FAILED, workflowEvents)
+	if !ok {
+		t.Fatal("Failed to find WFT-failed event in history")
+	}
+
+	c.testWFTFailedEvent(t, wft, event)
```
Comment (yycptt) [inline service/history/workflow/workflow_test/mutable_state_impl_test.go:326]: probably also test buffered signals really get flushed.

File: service/history/workflow/workflow_test/mutable_state_impl_test.go (hunk 1 above)
Comment (MichaelSnowden) [reply service/history/workflow/workflow_test/mutable_state_impl_test.go]: Done

File: service/history/workflow/workflow_test/mutable_state_impl_test.go (hunk 2 above)
Comment (MichaelSnowden) [reply service/history/workflow/workflow_test/mutable_state_impl_test.go]: Done

File: service/history/workflow/workflow_test/mutable_state_impl_test.go (hunk 3 above)
Comment (MichaelSnowden) [reply service/history/workflow/workflow_test/mutable_state_impl_test.go]: I think I just need to close one mutation transaction first? I'm not sure why, but I simplified this to just this:

```go
	_, err := ms.AddWorkflowExecutionStartedEvent(
		commonpb.WorkflowExecution{
			WorkflowId: "694B31C3-1FDC-4C9E-87BC-747D539BF0CD",
			RunId:      "3E1836B8-8692-440D-9495-45D9EABDED6B",
		},
		&historyservice.StartWorkflowExecutionRequest{
			Attempt:     1,
			NamespaceId: nsEntry.ID().String(),
			StartRequest: &workflowservice.StartWorkflowExecutionRequest{
				WorkflowType:        &commonpb.WorkflowType{Name: "workflow-type"},
				TaskQueue:           &taskqueuepb.TaskQueue{Name: "task-queue-name"},
				WorkflowRunTimeout:  timestamp.DurationPtr(200 * time.Second),
				WorkflowTaskTimeout: timestamp.DurationPtr(1 * time.Second),
			},
		},
	)
	if err != nil {
		t.Fatal(err)
	}

	_, _, err = ms.CloseTransactionAsMutation(c.transactionPolicy)
	if err != nil {
		t.Fatal(err)
	}
```

File: service/history/workflow/workflow_test/mutable_state_impl_test.go (hunk 4 above)
Comment (MichaelSnowden) [reply service/history/workflow/workflow_test/mutable_state_impl_test.go:312]: Added

File: service/history/workflow/workflow_test/mutable_state_impl_test.go [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+		t.Run(tc.name, tc.Run)
+	}
+}
+
+type mutationTestCase struct {
+	name              string
+	transactionPolicy workflow.TransactionPolicy
+	signals           int
+	maxNumEvents      int
+	expectFailure     bool
```
Comment (yycptt) [inline service/history/workflow/workflow_test/mutable_state_impl_test.go:92]: nit: maybe `expectFlush` ? Fail inflight workflow task is just how the first step of flush.

**Review by yycptt — APPROVED** [review]

File: service/history/workflow/workflow_test/mutable_state_impl_test.go [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+	for i := 0; i < c.signals; i++ {
+		c.addWorkflowExecutionSignaled(t, i, ms)
+	}
+
+	mu, workflowEvents, err := ms.CloseTransactionAsMutation(c.transactionPolicy)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_ = mu
```
[the remark this answers is not in the document]
Comment (MichaelSnowden) [reply service/history/workflow/workflow_test/mutable_state_impl_test.go]: Oops, thanks for catching that!

**Pull request merged**